pkg/tbtc/moved_funds_sweep.go

```diff
@@ -51,6 +80,183 @@ func (mfsp *MovedFundsSweepProposal) ValidityBlocks() uint64 {
 	return movedFundsSweepProposalValidityBlocks
 }
 
+type movedFundsSweepAction struct {
+	logger   *zap.SugaredLogger
+	chain    Chain
+	btcChain bitcoin.Chain
+
+	movedFundsSweepWallet wallet
+	transactionExecutor   *walletTransactionExecutor
+
+	proposal                     *MovedFundsSweepProposal
+	proposalProcessingStartBlock uint64
+	proposalExpiryBlock          uint64
+
+	signingTimeoutSafetyMarginBlocks uint64
+	broadcastTimeout                 time.Duration
+	broadcastCheckDelay              time.Duration
+}
+
+func newMovedFundsSweepAction(
+	logger *zap.SugaredLogger,
+	chain Chain,
+	btcChain bitcoin.Chain,
+	movedFundsSweepWallet wallet,
+	signingExecutor walletSigningExecutor,
+	proposal *MovedFundsSweepProposal,
+	proposalProcessingStartBlock uint64,
+	proposalExpiryBlock uint64,
+	waitForBlockFn waitForBlockFn,
+) *movedFundsSweepAction {
+	transactionExecutor := newWalletTransactionExecutor(
+		btcChain,
+		movedFundsSweepWallet,
+		signingExecutor,
+		waitForBlockFn,
+	)
+
+	return &movedFundsSweepAction{
+		logger:                           logger,
+		chain:                            chain,
+		btcChain:                         btcChain,
+		movedFundsSweepWallet:            movedFundsSweepWallet,
+		transactionExecutor:              transactionExecutor,
+		proposal:                         proposal,
+		proposalProcessingStartBlock:     proposalProcessingStartBlock,
+		proposalExpiryBlock:              proposalExpiryBlock,
+		signingTimeoutSafetyMarginBlocks: movedFundsSweepSigningTimeoutSafetyMarginBlocks,
+		broadcastTimeout:                 movedFundsSweepBroadcastTimeout,
+		broadcastCheckDelay:              movedFundsSweepBroadcastCheckDelay,
+	}
+}
+
+func (mfsa *movedFundsSweepAction) execute() error {
+	validateProposalLogger := mfsa.logger.With(
+		zap.String("step", "validateProposal"),
+	)
+
+	walletPublicKeyHash := bitcoin.PublicKeyHash(mfsa.wallet().publicKey)
+
+	err := ValidateMovedFundsSweepProposal(
+		validateProposalLogger,
+		walletPublicKeyHash,
+		mfsa.proposal,
+		mfsa.chain,
+	)
+	if err != nil {
+		return fmt.Errorf("validate proposal step failed: [%v]", err)
+	}
+
+	// Prepare the moved funds UTXO.
+	movingFundsTxHash := mfsa.proposal.MovingFundsTxHash
+	movingFundsTxOutputIndex := mfsa.proposal.MovingFundsTxOutputIndex
+
+	movingFundsTx, err := mfsa.btcChain.GetTransaction(
+		movingFundsTxHash,
+	)
+	if err != nil {
+		return fmt.Errorf("could not get moving funds transaction: [%v]", err)
+	}
+
+	movingFundsTxValue := movingFundsTx.Outputs[movingFundsTxOutputIndex].Value
+
+	movedFundsUtxo := &bitcoin.UnspentTransactionOutput{
+		Outpoint: &bitcoin.TransactionOutpoint{
+			TransactionHash: movingFundsTxHash,
+			OutputIndex:     movingFundsTxOutputIndex,
+		},
+		Value: movingFundsTxValue,
+	}
```

Comment:
Just for readability: Can we extract this code to a separate function?

Reply:
Done in daade95e83ccda1ccde914dc05bdd445c5f25509.